org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/NodeModelBuilder.java

```diff
@@ -203,13 +205,8 @@ public ICompositeNode compressAndReturnParent(ICompositeNode compositeNode) {
 				// if it refers not to a syntax error or a semantic object
 				EObject myGrammarElement = casted.getGrammarElement();
 				Object childGrammarElement = firstChild.basicGetGrammarElement();
-				EObject[] list = null;
-				if (childGrammarElement instanceof EObject) {
-					list = new EObject[] {myGrammarElement, (EObject) childGrammarElement};
-				} else {
-					list = newEObjectArray(myGrammarElement, (EObject[]) childGrammarElement);
-				}
-				casted.basicSetGrammarElement(cachedFoldedGrammarElements.intern(list));
+				EObject[] grammarElements = cachedFoldedGrammarElements.joinAndIntern(myGrammarElement, childGrammarElement);
```

Comment:
i wonder if there i a way to restructure this (or the old version)
in a way so that it could be customized in a subclass without 
package private troubles

Reply:
It wouldn't harm to introduce new protected methods / fields in this class.